Feat/improve readme (#27)
FIL-Builders/onramp-contracts: README.md

# On Ramp Contracts 🚀

Bringing decentralized storage to every blockchain! This project enables **dApps to store data on Filecoin** from **multiple L1/L2 networks** using cross-chain smart contracts.

## 📚 Table of Contents
- [Overview](#overview)

>   - [What is a Cross-Chain Data Bridge?](#what-is-a-cross-chain-data-bridge)
>   - [What are Onramp Contracts?](#what-are-onramp-contracts)

Reviewer: ```suggestion
```

> <h1 align="center">

Reviewer: Let's keep this aligned on the left.

Author: alright

  On Ramp Contracts 🚀
</h1>

Bringing decentralized storage to every blockchain! This project enables **dApps to store data on Filecoin** from **multiple L1/L2 networks** using cross-chain smart contracts.

## 📚 Table of Contents
- [Overview](#overview)
  - [What is a Cross-Chain Data Bridge?](#what-is-a-cross-chain-data-bridge)
  - [What are Onramp Contracts?](#what-are-onramp-contracts)
- [Architecture](#architecture)
- [Project Structure](#project-structure)
- [Getting Started](#getting-started)
  - [Prerequisites](#prerequisites)
  - [Installation Steps](#installation-steps)
- [Deployment Instructions](#deployment-instructions)
- [Configuration](#configuration)
- [Setting up xChain Client](#setting-up-xchain-client)
- [Usage](#usage)
- [Additional Resources](#additional-resources)
- [Contributing](#contributing)
- [License](#license)

## 🌍 Overview

> ### What is a Cross-Chain Data Bridge?
> A cross-chain data bridge allows applications on one blockchain (source chain) to interact with and store data on another blockchain (destination chain). In the context of Onramp Contracts, this means enabling any blockchain to leverage Filecoin's decentralized storage capabilities.
> 
> ### What are Onramp Contracts?
> Our smart contracts act as a **bridge** between various blockchains (like Linea, Avalanche, and Arbitrum) and **Filecoin** storage. The framework consists of:

Reviewer: ```suggestion
The cross-chain data bridge (aka, onramp contracts) allows applications on one blockchain (source chain) to store and retrieve data on the Filecoin decentralized network. This means that any blockchain can leverage Filecoin's decentralized storage capabilities for: 

- Cross-Chain Compatibility: Supports multiple L1 and L2 networks.
- Verifiable Data Storage: Utilizes Filecoin's robust storage network.
- Modular Architecture: Easily extendable to support additional networks and functionalities.

Our onramp contracts act as a **bridge** between various blockchains (like Linea, Avalanche, and Arbitrum) and **Filecoin** storage. The framework consists of:
```

✅ **Source Chains (L1/L2 networks)**

> - **`OnRampContract`** – Handles user deposits & cross-chain messaging

Reviewer: ```suggestion
- **`OnRampContract`** – Handles cross-chain storage requests & verification, and user payments.
```

✅ **Filecoin (Storage Destination)**

> - **`DealClientAxl`** – Receives data & initiates storage deals

Reviewer: ```suggestion
- **`DealClientAxl`** – Receives deal notification from Filecoin,  sends proof back to source chain. 
```

## Architecture

> The system works through three main components deployed across chains:
> 
> 1. **Source Chain Contracts**: 
>    - OnRampContract from `Onramp.sol`: Manages data submissions and payment
>    - AxelarBridge from `Oracle.sol`: Handles cross-chain communication
> 2. **Destination Chain Contract (Filecoin)**:
>    - DealClientAxl from `Prover-Axelar.sol`: Processes storage deals and manages data verification
> 3. **xChain Client**: Coordinates the data transfer and deal-making process

Reviewer: ```suggestion
The cross-chain data bridge works through two main components deployed across chains:

1. **Onramp Contracts**: 

- Source Chain Contracts: `Onramp.sol` & `Oracle.sol`
- Destination Chain Contract (Filecoin): `Prover.sol`

2. **[xChain Client](https://github.com/FIL-Builders/xchainClient)**: 
Monitoring storage requests from the source chain, aggregating data, and facilitating deal-making with storage providers or deal engines.
```

> ## Project Structure

Reviewer: Let's remove this section. Structure is going to be changed over time and it is not very helpful at this stage.

Author: Alright

```bash
npx hardhat deploy --tags Filecoin --network filecoin
```

### Step 2: Deploy Source Chain Contracts
🌉 Deploys `OnRampContract` & `AxelarBridge` on **your chosen L1/L2 source chain**.

**Example for Linea:**
```bash
npx hardhat deploy --tags SourceChain --network linea-sepolia
```

**Other supported networks:**
```bash
npx hardhat deploy --tags SourceChain --network arbitrum-sepolia
npx hardhat deploy --tags SourceChain --network avalanche
```

## 🔧 Configuration

### Step 3: Wire Filecoin with Source Chains
👀 **Automatically detects all deployed source chains** and configures **DealClientAxl**:
```bash
npx hardhat deploy --tags ConfigFilecoin --network filecoin
```

### Step 4: Configure Source Chains
🏗 **Sets up cross-chain messaging**:
```bash
npx hardhat deploy --tags ConfigSourceChain --network linea-sepolia
```

### Running Full Deployment in One Command
```bash
npx hardhat deploy --tags Filecoin --network filecoin && \
npx hardhat deploy --tags SourceChain --network linea-sepolia && \
npx hardhat deploy --tags ConfigFilecoin --network filecoin && \
npx hardhat deploy --tags ConfigSourceChain --network linea-sepolia
```

> ## 🛠 Setting Up xChain Client

Reviewer: ```suggestion
## **🛠 Setting Up the Off-Chain Components (xChain Client)**
The cross-chain data bridge requires running [xChain client](https://github.com/FIL-Builders/xchainClient) to process the storage requests and storages proof between the source chain and Filecoin.
```
